go/libraries/doltcore/doltdb/table_edit_session.go

```diff
@@ -0,0 +1,239 @@
+	}
+
+	fkCollection, err := root.GetForeignKeyCollection(ctx)
+	if err != nil {
+		return err
+	}
+	tes.root = root
+
+	for tableName, localTableEditor := range tes.tables {
+		t, ok, err := root.GetTable(ctx, tableName)
+		if err != nil {
+			return err
+		}
+		if !ok { // table was removed in newer root
+			delete(tes.tables, tableName)
+			continue
+		}
+		tSch, err := t.GetSchema(ctx)
+		if err != nil {
+			return err
+		}
+		newTableEditor, err := NewTableEditor(ctx, t, tSch)
+		if err != nil {
+			return err
+		}
+		localTableEditor.tableEditor = newTableEditor
+		localTableEditor.referencedTables, localTableEditor.referencingTables = fkCollection.KeysForTable(tableName)
+		err = tes.loadForeignKeys(ctx, localTableEditor)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// getTableEditor is the inner implementation for GetTableEditor, allowing recursive calls
+func (tes *TableEditSession) getTableEditor(ctx context.Context, tableName string, tableSch schema.Schema) (*SessionedTableEditor, error) {
+	if tes.root == nil {
+		return nil, fmt.Errorf("must call SetRoot before a table editor will be returned")
+	}
+
+	var t *Table
+	var err error
+	localTableEditor, ok := tes.tables[tableName]
+	if ok {
+		if tableSch == nil {
+			return localTableEditor, nil
+		} else if ok, err = schema.SchemasAreEqual(tableSch, localTableEditor.tableEditor.tSch); err == nil && ok {
+			return localTableEditor, nil
+		}
+		// Any existing references to this localTableEditor should be preserved, so we just change the underlying values
+		localTableEditor.referencedTables = nil
+		localTableEditor.referencingTables = nil
+	} else {
+		localTableEditor = &SessionedTableEditor{
+			tableEditSession:  tes,
+			tableEditor:       nil,
+			referencedTables:  nil,
+			referencingTables: nil,
+		}
+		tes.tables[tableName] = localTableEditor
+	}
+
+	t, ok, err = tes.root.GetTable(ctx, tableName)
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		return nil, fmt.Errorf("unable to create table editor as `%s` is missing", tableName)
+	}
+	if tableSch == nil {
+		tableSch, err = t.GetSchema(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	tableEditor, err := NewTableEditor(ctx, t, tableSch)
+	if err != nil {
+		return nil, err
+	}
+	localTableEditor.tableEditor = tableEditor
+	if tes.Props.ForeignKeyChecksDisabled {
+		return localTableEditor, nil
+	}
+
+	fkCollection, err := tes.root.GetForeignKeyCollection(ctx)
+	if err != nil {
+		return nil, err
+	}
+	localTableEditor.referencedTables, localTableEditor.referencingTables = fkCollection.KeysForTable(tableName)
+	err = tes.loadForeignKeys(ctx, localTableEditor)
+	if err != nil {
+		return nil, err
+	}
+
+	return localTableEditor, nil
+}
+
+// getTableEditorOrFail returns the table editor for this table or fails
+func (tes *TableEditSession) getTableEditorOrFail(tableName string) (*SessionedTableEditor, error) {
```

Comment:
So this function is directly due to feedback left last time. I had:
```go
referencedLte, ok := globalTableEditor.tables[foreignKey.ReferencedTableName]
```
To which your feedback was:

> Should call a method here, rather than accessing the field directly

This function is only called from a `SessionedTableEditor`, and it's only in the places where I was using the above code. No clients know of or even care that this function exists (it's lowercased to make it package-private).

I think there may be some confusion on the overall goal. So the design goal of `TableEditSession` and friends is to be a sort of drop-in replacement for just using a standard, non-foreign-key-aware `TableEditor`. Meaning a client, given a session, should just call `GetTableEditor`, use it like before, and call `GetRoot` when they're done. They should not worry about anything besides that.

Internally, `TableEditSession` uses one editor per table. A client does not need to know nor care about that. This gives advantages though, such as using internal table editors for value cascading from foreign keys and the such, and if an editor already exists due to the `TableEditSession` having already instantiated one for internal use, it can just return that already-existing one, and the logic stays the same.

Reply:
So calling a method rather than addressing the field is good. I just think it's the wrong method. If you want to do an existence check for a table's editor, make a method that does that. TableEditorExists or similar.

I don't remember the significance of the ok param in the previous revision. You sure it matters that a table editor already exists? Not obvious to me why that would be.